In [1]:
import pandas as pd
import numpy as np

# Soal 1

In [2]:
list_plat_mobil = ['B 1234 AB', 'B 6721 TY', 'B 1233 AY', 'B 6629 DD', 'B 1111 AM', 'B 6726 D', 
                   'D 11223 KJ', 'AE 44677 GH', 'AE 67269 AA']

genap = 0
ganjil = 0

for plat in list_plat_mobil:
    sisa = int(plat.split(' ')[1]) % 2
    if sisa == 0:
        genap += 1
    else:
        ganjil += 1

print(f'Jumlah plat genap: {genap}\nJumlah plat ganjil: {ganjil}')

Jumlah plat genap: 2
Jumlah plat ganjil: 7


# Soal 2

Buatlah deret fibonacci 12 angka yang dimulai dari 2 dan 3 menggunakan iterasi cara for atau while.

Output yang diharapkan adalah [2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377] (5 poin)

In [3]:
jumlah = 12
deret = [2,3]

while len(deret) < jumlah:
    deret.append(deret[-1] + deret[-2])
    
print(deret)

[2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377]


# Soal 3

In [4]:
# combination with replacement misal: (c1,c1,c2) boleh

In [5]:
list_id_category = ['c1', 'c2', 'c3', 'c4', 'c5', 'c6', 'c7' , 'c8', 'c9', 'c10']
bundling = []

for i in list_id_category:
    for j in list_id_category:
        for k in list_id_category:
            if i == j == k: 
                pass # tidak boleh hanya 1 barang
            else:
                temp = [i,j,k]
                temp.sort()
                if temp not in bundling:
                    bundling.append(temp)
            
len(bundling)

210

# Soal 4

Buatlah sebuah function string_processor dengan inputan 2 parameter, yaitu suatu string 1 kata dan nama proses yang akan dilakukan. Function ini memiliki 4 proses

In [6]:
def string_processor(kata, proses='first_char'):
    if proses == 'first_char':
        return kata[0]
    elif proses == 'last_char':
        return kata[-1]
    elif proses == 'num_chars':
        return len(''.join(kata.split()))
    elif proses == 'num_vowels':
        kata_vowel = [huruf for huruf in kata if huruf.lower() in 'aiueo']
        return len(kata_vowel)
    else:
        print('input proses salah')

In [7]:
string_processor('''TugAs ProgrAm
                    mIng Pyth         On''','first_char')

'T'

In [8]:
string_processor('''TugAs ProgrAm
                    mIng Pyth         On''','last_char')

'n'

In [9]:
string_processor('''TugAs ProgrAm
                    mIng Pyth         On''','num_chars')
# disini whitespace tidak dihitung mas, 5+7+4+4+2 = 22

22

In [10]:
string_processor('''TugAs ProgrAm
                    mIng Pyth         On''','num_vowels')
# uAoAIO = 6

6

# Soal 5

In [11]:
df = pd.read_csv('rakamin_store_city_monthly.csv', delimiter=';')
df.head()

month  user_id                 username  \
0  2016-01       16            Kevin Ningsih   
1  2016-01       25            Sakti Hasaniy   
2  2016-01       39              Izhar Halim   
3  2016-02        8  Dikposa Yosua Krisnanto   
4  2016-02       34          Ismail Wardhani   

                               email             kota  total_visit  \
0         kevinningsih@roketmail.com  Jakarta Selatan           95   
1           saktihasaniy@outlook.com    Jakarta Pusat           97   
2               izharhalim@gmail.com    Jakarta Timur           84   
3  dikposayosuakrisnanto@hotmail.com    Jakarta Pusat          113   
4         ismailwardhani@hotmail.com  Jakarta Selatan           25   

   total_order  total_amount  
0           15         45724  
1            1         26665  
2           33        899001  
3           11         22012  
4            2         55620

In [12]:
df['is_gmail'] = df.apply(lambda x: '@gmail' in x['email'].lower(), axis=1)
df['year'] = df['month'].apply(lambda x: x.split('-')[0])
df.head()

month  user_id                 username  \
0  2016-01       16            Kevin Ningsih   
1  2016-01       25            Sakti Hasaniy   
2  2016-01       39              Izhar Halim   
3  2016-02        8  Dikposa Yosua Krisnanto   
4  2016-02       34          Ismail Wardhani   

                               email             kota  total_visit  \
0         kevinningsih@roketmail.com  Jakarta Selatan           95   
1           saktihasaniy@outlook.com    Jakarta Pusat           97   
2               izharhalim@gmail.com    Jakarta Timur           84   
3  dikposayosuakrisnanto@hotmail.com    Jakarta Pusat          113   
4         ismailwardhani@hotmail.com  Jakarta Selatan           25   

   total_order  total_amount  is_gmail  year  
0           15         45724     False  2016  
1            1         26665     False  2016  
2           33        899001      True  2016  
3           11         22012     False  2016  
4            2         55620     False  2016

In [13]:
df[(df['year'] == '2018')].groupby('is_gmail').agg({'user_id' : 'nunique'}).reset_index()

is_gmail  user_id
0     False       14
1      True        3

Di tahun 2018 ada 3 user yang pakai email gmail (is_gmail = True)

# Soal 6

In [14]:
df_superstore = pd.read_excel('superstore.xls')

In [15]:
df_superstore.head()

Row ID        Order ID Order Date  Ship Date       Ship Mode Customer ID  \
0       1  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
1       2  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
2       3  CA-2016-138688 2016-06-12 2016-06-16    Second Class    DV-13045   
3       4  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   
4       5  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   

     Customer Name    Segment        Country             City  ...  \
0      Claire Gute   Consumer  United States        Henderson  ...   
1      Claire Gute   Consumer  United States        Henderson  ...   
2  Darrin Van Huff  Corporate  United States      Los Angeles  ...   
3   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   
4   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   

  Postal Code  Region       Product ID         Category Sub-Category  \
0       42420   South  FUR-BO-10001798        Furniture    Bookcases   
1       42420   South  FUR-CH-10000454        Furniture       Chairs   
2       90036    West  OFF-LA-10000240  Office Supplies       Labels   
3       33311   South  FUR-TA-10000577        Furniture       Tables   
4       33311   South  OFF-ST-10000760  Office Supplies      Storage   

                                        Product Name     Sales  Quantity  \
0                  Bush Somerset Collection Bookcase  261.9600         2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400         3   
2  Self-Adhesive Address Labels for Typewriters b...   14.6200         2   
3      Bretford CR4500 Series Slim Rectangular Table  957.5775         5   
4                     Eldon Fold 'N Roll Cart System   22.3680         2   

   Discount    Profit  
0      0.00   41.9136  
1      0.00  219.5820  
2      0.00    6.8714  
3      0.45 -383.0310  
4      0.20    2.5164  

[5 rows x 21 columns]

In [16]:
# Ini function hitung salesnya, dengan parameter default 2017
def hitung_sales(region, tahun=2017, df=df_superstore):
    df['order_year'] = df['Order Date'].dt.year
    filter_region = df['Region'] == region
    filter_tahun = df['order_year'] == tahun
    return df[filter_region & filter_tahun].groupby('order_year').agg({'Sales': 'sum',
                                                                       'Profit': 'sum'
                                                                      }).reset_index()

In [17]:
hitung_sales('South', 2016)

order_year       Sales      Profit
0        2016  93610.2235  17702.8084

In [18]:
hitung_sales('West')

order_year        Sales      Profit
0        2017  250128.3655  43808.9561

# Soal 7

In [19]:
# data dikelompokkan berdasarkan customer ID -> karena ada 1 customer berulang kali order
# Hati-hati, karena ini segmentasi user maka perlu melakukan agregasi di level user terlebih dulu sebelum melakukan segmentasi
df_customer = df_superstore[['Customer ID','Sales', 'Quantity', 'Discount', 'Profit']].groupby('Customer ID').agg({'Sales': 'sum',
                                                                                                                   'Quantity': 'sum',
                                                                                                                   'Discount': 'mean',
                                                                                                                   'Profit': 'sum'
                                                                                                                  }).reset_index()

df_customer.head()

Customer ID     Sales  Quantity  Discount    Profit
0    AA-10315  5563.560        30  0.090909 -362.8825
1    AA-10375  1056.390        41  0.080000  277.3824
2    AA-10480  1790.512        36  0.016667  435.8274
3    AA-10645  5086.935        64  0.063889  857.8033
4    AB-10015   886.156        13  0.066667  129.3465

In [20]:
df_customer.describe().T

count         mean          std        min          25%        50%  \
Sales     793.0  2896.848500  2628.670117     4.8330  1146.050000  2256.3940   
Quantity  793.0    47.759142    24.842915     2.0000    30.000000    44.0000   
Discount  793.0     0.157482     0.089071     0.0000     0.090909     0.1500   
Profit    793.0   361.156396   894.261812 -6626.3895    36.613100   227.8338   

                  75%         max  
Sales     3785.276000  25043.0500  
Quantity    63.000000    150.0000  
Discount     0.211111      0.7000  
Profit     560.007800   8981.3239

*Buatlah 3 kelompok 'segment' (high, medium, dan low) berdasarkan performance Sales*

3 kelompok segmen dibuat berdasarkan quartile. Alasannya supaya pembagian tiap kelompok dapat terukur berapa kira-kira yang masuk kelompok tersebut.

* Q1 ke bawah = 25% pengguna (Low)
* Q1-Q3 = 50% pengguna (Medium/average)
* Q4 ke atas = 25% pengguna (High)

Apabila dilihat dari sales, maka:
* Low = sales kurang dari 1146.05
* Medium = sales di antara 1146.05 - 3785.276
* High = sales lebih dari 3785.276

In [21]:
# ini fungsi segmentasinya
def segmentasi(df):
    Q1 = df_customer['Sales'].quantile(q = 0.25)
    Q3 = df_customer['Sales'].quantile(q = 0.75)
    try:
        if df['Sales'] < Q1:
            return 'Low'
        elif df['Sales'] > Q3:
            return 'High'
        else:
            return 'Medium'
    except:
        return 'Other'

In [22]:
# Tentukan jumlah customer pada masing-masing segment di tahun 2017!

# fungsi di bawah ini digunakan untuk mengelompokkan pembelian user per-tahun (groupby order_year, customer ID), jadi bisa difilter total pembelian user dalam tahun tertentu.
def kelompokkan_tahun(df=df_superstore):
    df['order_year'] = df['Order Date'].dt.year
    return df[['order_year','Customer ID','Sales', 'Quantity', 'Discount', 'Profit']].groupby(['order_year','Customer ID']).agg({'Sales': 'sum',
                                                                                                                                 'Quantity': 'sum',
                                                                                                                                 'Discount': 'mean',
                                                                                                                                 'Profit': 'sum'
                                                                                                                                }).reset_index()

df_cust_year = kelompokkan_tahun()
df_cust_year.head()

order_year Customer ID     Sales  Quantity  Discount    Profit
0        2014    AA-10315   756.048         9  0.050000  280.7050
1        2014    AA-10375    50.792         8  0.200000   16.7139
2        2014    AA-10480    27.460         2  0.000000    9.8856
3        2014    AA-10645  1434.330        18  0.166667   77.6417
4        2014    AB-10015   322.216         9  0.100000   12.4785

In [23]:
# menggunakan apply lambda untuk membuat baris segmentasi
df_cust_year['segmentasi'] = df_cust_year.apply(lambda x: segmentasi(x), axis = 1)

In [24]:
# anyway, profit tidak sebanding dengan sales. Tapi karena segmentasi disuruh berdasarkan sales, jadi hasilnya ada yang profit tinggi tapi Low, profit rendah tapi medium.
df_cust_year.head()

order_year Customer ID     Sales  Quantity  Discount    Profit segmentasi
0        2014    AA-10315   756.048         9  0.050000  280.7050        Low
1        2014    AA-10375    50.792         8  0.200000   16.7139        Low
2        2014    AA-10480    27.460         2  0.000000    9.8856        Low
3        2014    AA-10645  1434.330        18  0.166667   77.6417     Medium
4        2014    AB-10015   322.216         9  0.100000   12.4785        Low

In [25]:
# di groupby berdasarkan segmentasinya untuk mengetahui jumlah user (nunique customer ID) dalam tahun 2017 (filter)
df_cust_year[df_cust_year['order_year'] == 2017].groupby('segmentasi').agg({'Customer ID': 'nunique'}).reset_index()

segmentasi  Customer ID
0       High           34
1        Low          486
2     Medium          173

# Soal 8

In [26]:
# RFM

In [27]:
df_superstore.head()

Row ID        Order ID Order Date  Ship Date       Ship Mode Customer ID  \
0       1  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
1       2  CA-2016-152156 2016-11-08 2016-11-11    Second Class    CG-12520   
2       3  CA-2016-138688 2016-06-12 2016-06-16    Second Class    DV-13045   
3       4  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   
4       5  US-2015-108966 2015-10-11 2015-10-18  Standard Class    SO-20335   

     Customer Name    Segment        Country             City  ... Region  \
0      Claire Gute   Consumer  United States        Henderson  ...  South   
1      Claire Gute   Consumer  United States        Henderson  ...  South   
2  Darrin Van Huff  Corporate  United States      Los Angeles  ...   West   
3   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...  South   
4   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...  South   

        Product ID         Category Sub-Category  \
0  FUR-BO-10001798        Furniture    Bookcases   
1  FUR-CH-10000454        Furniture       Chairs   
2  OFF-LA-10000240  Office Supplies       Labels   
3  FUR-TA-10000577        Furniture       Tables   
4  OFF-ST-10000760  Office Supplies      Storage   

                                        Product Name     Sales Quantity  \
0                  Bush Somerset Collection Bookcase  261.9600        2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400        3   
2  Self-Adhesive Address Labels for Typewriters b...   14.6200        2   
3      Bretford CR4500 Series Slim Rectangular Table  957.5775        5   
4                     Eldon Fold 'N Roll Cart System   22.3680        2   

   Discount    Profit  order_year  
0      0.00   41.9136        2016  
1      0.00  219.5820        2016  
2      0.00    6.8714        2016  
3      0.45 -383.0310        2015  
4      0.20    2.5164        2015  

[5 rows x 22 columns]

In [28]:
df_ss = df_superstore[['Order Date', 'Customer ID', 'Sales', 'Quantity']]
df_ss['recency'] = df_ss['Order Date'].max() - df_ss['Order Date']
df_ss['monetary'] = df_ss['Quantity'] * df_ss['Sales']
df_ss.head()

C:\Users\Asif\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
C:\Users\Asif\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


Order Date Customer ID     Sales  Quantity  recency   monetary
0 2016-11-08    CG-12520  261.9600         2 417 days   523.9200
1 2016-11-08    CG-12520  731.9400         3 417 days  2195.8200
2 2016-06-12    DV-13045   14.6200         2 566 days    29.2400
3 2015-10-11    SO-20335  957.5775         5 811 days  4787.8875
4 2015-10-11    SO-20335   22.3680         2 811 days    44.7360

In [29]:
df_rfm = df_ss.groupby('Customer ID').agg({'recency':'min',
                                          'Order Date': 'count',
                                          'monetary': 'sum'
                                         }).reset_index()
df_rfm.columns = ['Customer ID', 'recency', 'frequency', 'monetary']
df_rfm.head()

Customer ID  recency  frequency   monetary
0    AA-10315 184 days         11  16073.344
1    AA-10375  19 days         15   2734.830
2    AA-10480 259 days         12   5493.936
3    AA-10645  55 days         18  22927.855
4    AB-10015 415 days          6   1984.480

In [30]:
df = df_rfm.copy()

In [31]:
df['r_rank'] = 5 - pd.qcut(df['recency'], q=5, labels = False)
df['f_rank'] = pd.qcut(df['frequency'], q=5, labels = False) + 1
df['m_rank'] = pd.qcut(df['monetary'], q=5, labels = False) + 1

df.head()

Customer ID  recency  frequency   monetary  r_rank  f_rank  m_rank
0    AA-10315 184 days         11  16073.344       2       3       4
1    AA-10375  19 days         15   2734.830       5       4       1
2    AA-10480 259 days         12   5493.936       1       3       2
3    AA-10645  55 days         18  22927.855       3       4       5
4    AB-10015 415 days          6   1984.480       1       1       1

In [32]:
r = 'recency'
f = 'frequency'
m = 'monetary'
weight=[1/3,1/3,1/3]

In [33]:
df['rfm_score'] = np.average(df[['r_rank','f_rank', 'm_rank']], weights=weight, axis=1)

* rfm score >4.5 : Top Customer
* 4.5 > rfm score > 4 : High Value Customer
* 4>rfm score >3 : Medium value customer
* 3>rfm score>1.6 : Low-value customer
* rfm score<1.6 :Lost Customer

In [34]:
bins = [0, 1.6, 3, 4, 4.5, 5]
names = ['Lost Customer', 'Low-value Customer', 'Medium-value Customer', 'High-value Customer', 'Top Customer']
df['segment'] = pd.cut(df['rfm_score'], bins, labels=names)
df.head()

Customer ID  recency  frequency   monetary  r_rank  f_rank  m_rank  \
0    AA-10315 184 days         11  16073.344       2       3       4   
1    AA-10375  19 days         15   2734.830       5       4       1   
2    AA-10480 259 days         12   5493.936       1       3       2   
3    AA-10645  55 days         18  22927.855       3       4       5   
4    AB-10015 415 days          6   1984.480       1       1       1   

   rfm_score                segment  
0   3.000000     Low-value Customer  
1   3.333333  Medium-value Customer  
2   2.000000     Low-value Customer  
3   4.000000  Medium-value Customer  
4   1.000000          Lost Customer

In [35]:
df = df_rfm.copy()

df['r_rank'] = 5 - pd.qcut(df[r], q=5, labels = False)
df['f_rank'] = pd.qcut(df[f], q=5, labels = False) + 1
df['m_rank'] = pd.qcut(df[m], q=5, labels = False) + 1

In [36]:
fr_grid = [
    ['Hibernating', 'Hibernating', 'About To Sleep', 'Promising', 'New customers'],
    ['Hibernating', 'Hibernating', 'About To Sleep', 'Potential Loyalist', 'Potential Loyalist'],
    ['At Risk', 'At Risk', 'Need Attention', 'Potential Loyalist', 'Potential Loyalist'],
    ['At Risk', 'At Risk', 'Loyal Customers', 'Loyal Customers', 'Champions'],
    ["Can't Lose Them", "Can't Lose Them", 'Loyal Customers', 'Loyal Customers', 'Champions']
]

In [37]:
df[['f_rank', 'r_rank']].apply(lambda x: fr_grid[x.f_rank - 1][x.r_rank -1], axis = 1)

0                 At Risk
1               Champions
2                 At Risk
3         Loyal Customers
4             Hibernating
              ...        
788       Loyal Customers
789    Potential Loyalist
790    Potential Loyalist
791       Loyal Customers
792           Hibernating
Length: 793, dtype: object

## Result no 8

In [38]:
import pandas as pd
import numpy as np
from datetime import date

def rfm_analysis(df,r,f,m,method='simple', weight=[1/3,1/3,1/3],grid='fr'):
    """
    Objective:
    Consumer segmentation based on a simple RFM analysis
    
    Arguments:
    - df = dataframe to segment, type pandas.DataFrame
    - r = column name for recency data, type string
    - f = column name for frequency data, type string
    - m = column name for monetary data, type string
    - method = {'simple' or 'fr-grid'}, default 'simple'
    - weight = list containing weight of each rfm variables, used if method = 'simple'
        expected value = [r_weight, f_weight, m_weight], default [1/3,1/3,1/3]
    
    Return: Pandas.DataFrame containing segmentation label
    """
    
    try:
        df = df.copy()
        df['r_rank'] = 5 - pd.qcut(df[r], q=5, labels = False)
        df['f_rank'] = pd.qcut(df[f], q=5, labels = False) + 1
        df['m_rank'] = pd.qcut(df[m], q=5, labels = False) + 1
        if(method == 'simple'):
            if(sum(weight) != 1):
                raise ValueError('Jumlah variabel weight harus 1')
            else:
                df['rfm_score'] = np.average(df[['r_rank','f_rank', 'm_rank']], weights=weight, axis=1)
                bins = [0, 1.6, 3, 4, 4.5, 5]
                names = ['Lost Customer', 'Low-value Customer', 'Medium-value Customer', 'High-value Customer', 'Top Customer']
                df['segment'] = pd.cut(df['rfm_score'], bins, labels=names)
                return df
        elif(method == 'fr-grid'):
            fr_grid = [
                ['Hibernating', 'Hibernating', 'About To Sleep', 'Promising', 'New customers'],
                ['Hibernating', 'Hibernating', 'About To Sleep', 'Potential Loyalist', 'Potential Loyalist'],
                ['At Risk', 'At Risk', 'Need Attention', 'Potential Loyalist', 'Potential Loyalist'],
                ['At Risk', 'At Risk', 'Loyal Customers', 'Loyal Customers', 'Champions'],
                ["Can't Lose Them", "Can't Lose Them", 'Loyal Customers', 'Loyal Customers', 'Champions']
            ] # sometimes the dumbest method win.
            df['segment'] = df[['f_rank', 'r_rank']].apply(lambda x: fr_grid[x.f_rank - 1][x.r_rank - 1], axis = 1)
            return df
    except:
        print('Something went wrong')

In [39]:
df = df_rfm.copy()

In [40]:
rfm_analysis(df = df, r = 'recency', f = 'frequency', m = 'monetary', method = 'simple', weight = [1/4,1/4,2/4]).head()

Customer ID  recency  frequency   monetary  r_rank  f_rank  m_rank  \
0    AA-10315 184 days         11  16073.344       2       3       4   
1    AA-10375  19 days         15   2734.830       5       4       1   
2    AA-10480 259 days         12   5493.936       1       3       2   
3    AA-10645  55 days         18  22927.855       3       4       5   
4    AB-10015 415 days          6   1984.480       1       1       1   

   rfm_score                segment  
0       3.25  Medium-value Customer  
1       2.75     Low-value Customer  
2       2.00     Low-value Customer  
3       4.25    High-value Customer  
4       1.00          Lost Customer

In [41]:
rfm_analysis(df = df, r = 'recency', f = 'frequency', m = 'monetary', method = 'fr-grid').head()

Customer ID  recency  frequency   monetary  r_rank  f_rank  m_rank  \
0    AA-10315 184 days         11  16073.344       2       3       4   
1    AA-10375  19 days         15   2734.830       5       4       1   
2    AA-10480 259 days         12   5493.936       1       3       2   
3    AA-10645  55 days         18  22927.855       3       4       5   
4    AB-10015 415 days          6   1984.480       1       1       1   

           segment  
0          At Risk  
1        Champions  
2          At Risk  
3  Loyal Customers  
4      Hibernating